### PR TITLE
[Ursa Log Analyzer] Generate the tags and schema handler for proto struct

### PR DESCRIPTION
**Pull request by wenrui-meng-ai**
### Description

- Generate the go struct tag based on the mapping between go type and proto type to the parquet type
- Generate schema handler with the generated tags
- Special handling for the Enum(convert to string) and Timestamp(convert to int64 mills) proto specific use 
- Fix a few minor issues in the recursive handling slice 

[Log Analyzer Framework](https://docs.google.com/document/d/1dZGKIlg3sqaNgDupLJ1tiroQvEoLuZXxg81GFJ16vYI/edit)

### How this code was tested

unit test.
`make test`

### Changelog

+no-changelog

### Issues fixed (optional)
close #182667

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -3,3 +3,5 @@ example/output/*
 
 # exception to the rule
 !example/output/.gitkeep
+
+.vscode/*
```
Comment (samserpoosh) [inline .gitignore:7]: We don't have this already? 😲 Looks like it should already [exists](https://github.com/AppliedIntuition/applied2/blob/0020aa7df58820227e0c3870103d0797e065d4ef/.gitignore#L80).

File: schema/proto.go [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+
+const (
+	ParquetFieldName = "name"
+
+	ParquetType               = "type"
+	ParquetConvertedType      = "convertedtype"
+	ParquetValueType          = "valuetype"
+	ParquetKeyType            = "keytype"
+	ParquetKeyConvertedType   = "keyconvertedtype"
+	ParquetValueConvertedType = "valueconvertedtype"
```
Comment (samserpoosh) [inline schema/proto.go:22]: It would be nice to either include one sentence description on the purpose of each item here or link to the docs for those who are less familiar with parquet and its schematic aspects.

File: schema/proto.go [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+func IsPrimitiveParquetType(tp string) bool {
+	switch tp {
+	case ParquetTypeBoolean,
+		ParquetTypeInt32,
+		ParquetTypeInt64,
+		ParquetTypeInt96,
+		ParquetTypeFloat,
+		ParquetTypeDouble,
+		ParquetTypeByteArray,
+		ParquetTypeFixedLenByteArray:
```
Comment (samserpoosh) [inline schema/proto.go:76]: We consider **arrays** as primitive here?

File: schema/proto.go [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+	ConvertedTypeBson            = "BSON"
+	ConvertedTypeInterval        = "INTERVAL"
+
+	ProtoEnumMethodName   = "Enum"
+	ProtoStringMethodName = "String"
+)
+
+var ProtoTimestampType = reflect.TypeOf(timestamppb.Timestamp{})
+
+func IsPrimitiveParquetType(tp string) bool {
```
Comment (samserpoosh) [inline schema/proto.go]: NIT: IF `tp` is meant to represent "type", let's rename it to `type` or `typ` (in case "type" is a reserved word in Go). It's more expressive that way IMHO.

File: schema/proto.go [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+	switch name {
+	case "state", "sizeCache", "unknownFields":
+		return true
+	default:
+		return false
+
+	}
+}
+
+func GetDefinedTypeTag(tp reflect.Type) (tags map[string]string, err error) {
```
Comment (samserpoosh) [inline schema/proto.go]:
```suggestion
func GetDefinedTypeTag(type reflect.Type) (tags map[string]string, err error) {
```

File: .gitignore (hunk 1 above)
Comment (samserpoosh) [reply .gitignore:7]: Synced live ... This is a FORK from an open-source repository and **not** our `applied2` 😄 We're good here.

File: schema/proto.go (hunk 5 above)
Comment (samserpoosh) [inline schema/proto.go]: Please add a description above this method explaining the performance rationale behind using `switch/case` vs a `map` as we discussed live. Thanks!

File: schema/proto.go [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+	case reflect.Int8:
+		tags[ParquetType] = ParquetTypeInt32
+		tags[ParquetConvertedType] = ConvertedTypeInt8
+	case reflect.Int16:
+		tags[ParquetType] = ParquetTypeInt32
+		tags[ParquetConvertedType] = ConvertedTypeInt16
+	case reflect.Int32:
+		// speical handling for the proto enum generated constant, it checks Enum() method generated. It's not ideal but no better way to do it.
+		if _, exists := tp.MethodByName(ProtoEnumMethodName); exists {
+			tags[ParquetType] = ParquetTypeByteArray
```
Comment (samserpoosh) [inline schema/proto.go:146]: `ParquetTypeByteArray` is how we represent **string** in Parquet's Type System, right?

File: schema/proto.go [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+	ConvertedTypeMap             = "MAP"
+	ConvertedTypeList            = "LIST"
+	ConvertedTypeEnum            = "ENUM"
+	ConvertedTypeDecimal         = "DECIMAL"
+	ConvertedTypeDate            = "DATE"
+	ConvertedTypeTimeMills       = "TIME_MILLIS"
+	ConvertedTypeTimeMicros      = "TIME_MICROS"
+	ConvertedTypeTimestampMills  = "TIMESTAMP_MILLIS"
+	ConvertedTypeTimestampMicros = "TIMESTAMP_MICROS"
+	ConvertedTypeUnit8           = "UINT_8"
```
Comment (samserpoosh) [inline schema/proto.go]:
```suggestion
	ConvertedTypeUnint8           = "UINT_8"
```

File: schema/proto.go [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+		} else {
+			tags[ParquetType] = ParquetTypeInt32
+		}
+	case reflect.Int64:
+		tags[ParquetType] = ParquetTypeInt64
+	case reflect.Int:
+		tags[ParquetType] = ParquetTypeInt64
+	case reflect.Uint8:
+		tags[ParquetType] = ParquetTypeInt32
+		tags[ParquetConvertedType] = ConvertedTypeUnit8
```
Comment (samserpoosh) [inline schema/proto.go]:
```suggestion
		tags[ParquetConvertedType] = ConvertedTypeUnint8
```

File: schema/proto.go [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+	case reflect.String:
+		tags[ParquetType] = ParquetTypeByteArray
+	case reflect.Pointer, reflect.UnsafePointer, reflect.Uintptr, reflect.Interface:
+		tags, err = GetDefinedTypeTag(tp.Elem())
+	case reflect.Array, reflect.Slice:
+		tags, err = getListTag(tp)
+	case reflect.Map:
+		tags, err = getMapTag(tp)
+	case reflect.Struct:
+		// do nothing since the struct tag is not needed in schema generation
```
Comment (samserpoosh) [inline schema/proto.go:183]: Why is that?

File: schema/proto.go [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+	default:
+		return nil, fmt.Errorf("Type %s is not supported in generating tags.", tp)
```
Comment (samserpoosh) [inline schema/proto.go]:
```suggestion
		return nil, fmt.Errorf("Type '%s' is not supported in generating tags.", tp)
```

Comment (samserpoosh) [reply schema/proto.go]: This ^^ is my OCD and it'll make it easier to read/highlight the embedded value in logging outcome usually, lol

File: schema/proto.go [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+	}
+	return
```
Comment (samserpoosh) [inline schema/proto.go:188]: Should this be `return tags, nil` instead?

File: schema/proto.go [hunk 12]
```diff
@@ -0,0 +1,386 @@
...
+	case reflect.Float64:
+		tags[ParquetType] = ParquetTypeDouble
+	case reflect.String:
+		tags[ParquetType] = ParquetTypeByteArray
+	case reflect.Pointer, reflect.UnsafePointer, reflect.Uintptr, reflect.Interface:
+		tags, err = GetDefinedTypeTag(tp.Elem())
+	case reflect.Array, reflect.Slice:
+		tags, err = getListTag(tp)
+	case reflect.Map:
+		tags, err = getMapTag(tp)
```
Comment (samserpoosh) [inline schema/proto.go]: In case these `GetDefinedTypeTag`, `getListTag` or `getMapTag` calls return an **error**, should we then short-circuit and return an ERROR in the initial method as well? Or we wanna skip the problematic column and move on gracefully?

File: schema/proto.go [hunk 13]
```diff
@@ -0,0 +1,386 @@
...
+	case reflect.Struct:
+		// do nothing since the struct tag is not needed in schema generation
+	default:
+		return nil, fmt.Errorf("Type %s is not supported in generating tags.", tp)
+	}
+	return
+}
+
+// Generate the tag for the struct field when there is no predefined tag
+func GenerateFieldTag(field reflect.StructField) (tagString string, err error) {
```
Comment (samserpoosh) [inline schema/proto.go]:
```suggestion
func GenerateFieldTag(field reflect.StructField) (tag string, err error) {
```

File: schema/proto.go [hunk 14]
```diff
@@ -0,0 +1,386 @@
...
+	tags[ParquetFieldName] = field.Name
+	return getTagStringFromTags(tags), nil
+}
+
+func getTagStringFromTags(tags map[string]string) string {
+	var pairs []string
+	for key, value := range tags {
+		pairs = append(pairs, fmt.Sprintf("%s=%s", key, value))
+	}
+	return strings.Join(pairs, ", ")
```
Comment (samserpoosh) [inline schema/proto.go:206]: So this would output something like the following I imagine?

```
name=foo,type=INT_32,convertedtype=INT_8,...
```

File: schema/proto.go [hunk 15]
```diff
@@ -0,0 +1,386 @@
...
+		tags[ParquetType] = ParquetTypeInt64
+		tags[ParquetConvertedType] = ConvertedTypeUnint64
+	case reflect.Float32:
+		tags[ParquetType] = ParquetTypeFloat
+	case reflect.Float64:
+		tags[ParquetType] = ParquetTypeDouble
+	case reflect.String:
+		tags[ParquetType] = ParquetTypeByteArray
+	case reflect.Pointer, reflect.UnsafePointer, reflect.Uintptr, reflect.Interface:
+		tags, err = GetDefinedTypeTag(tp.Elem())
```
Comment (samserpoosh) [inline schema/proto.go]: How would this work with an `Interface`? It's not going to lead to a `StructField` representing that Interface and its fields? Can you maybe provide an example?

File: schema/proto.go [hunk 16]
```diff
@@ -0,0 +1,386 @@
...
+	for key, value := range tags {
+		pairs = append(pairs, fmt.Sprintf("%s=%s", key, value))
+	}
+	return strings.Join(pairs, ", ")
+}
+
+func getMapTag(tp reflect.Type) (tags map[string]string, err error) {
+	tags = make(map[string]string)
+	tags[ParquetType] = ParquetTypeMap
+	tags[ParquetConvertedType] = ParquetTypeMap
```
Comment (samserpoosh) [inline schema/proto.go]: I thought we'd only need both `ParquetType` and `ParquetConvertedType` when they are **not** the same. But looks like here they're both the **same** values 🤔

File: schema/proto.go [hunk 17]
```diff
@@ -0,0 +1,386 @@
...
+			return nil, elementErr
+		}
+		if elementType, exists := elementTags[ParquetType]; exists && IsPrimitiveParquetType(elementType) {
+			tags[ParquetValueType] = elementType
+			if elementConvertedType, exists := elementTags[ParquetConvertedType]; exists {
+				tags[ParquetValueConvertedType] = elementConvertedType
+			}
+		}
+	}
+	return
```
Comment (samserpoosh) [inline schema/proto.go]: Same question here ... The return type appears to be `(tags map[string]string, err error)`. Does that mean this should be changed to:

```go
return tags, nil
```

File: schema/proto.go [hunk 18]
```diff
@@ -0,0 +1,386 @@
...
+		if elementErr != nil {
+			return nil, elementErr
+		}
+		if elementType, exists := elementTags[ParquetType]; exists && IsPrimitiveParquetType(elementType) {
+			tags[ParquetValueType] = elementType
+			if elementConvertedType, exists := elementTags[ParquetConvertedType]; exists {
+				tags[ParquetValueConvertedType] = elementConvertedType
+			}
+		}
+	}
```
Comment (samserpoosh) [inline schema/proto.go:239]: NIT: These `if IsPrimitiveOrPointerGoKind(tp.Key().Kind())` and `if IsPrimitiveOrPointerGoKind(tp.Elem().Kind())` blocks have a 90%+ overlap, maybe we can extract into a nice utility method and pass in things like `ParquetValueType`, `ParquetValueConvertedType` and their "xxzKeyxxx" counterparts and DRY this up a bit.

It might be an overkill so ignore this comment in case you think so 😄

File: schema/proto.go [hunk 19]
```diff
@@ -0,0 +1,386 @@
...
+	return
```
Comment (samserpoosh) [inline schema/proto.go]:
```suggestion
	return tags, nil
```

Maybe?

File: schema/proto.go [hunk 20]
```diff
@@ -0,0 +1,386 @@
...
+}
+
+func getListTag(tp reflect.Type) (tags map[string]string, err error) {
+	tags = make(map[string]string)
+	tags[ParquetType] = ParquetTypeList
+	tags[ParquetConvertedType] = ParquetTypeList
```
Comment (samserpoosh) [inline schema/proto.go]: Same question about having the identical value in both `ParquetType` and `ParquetConvertedType` cases.

File: schema/proto.go [hunk 21]
```diff
@@ -0,0 +1,386 @@
...
+			tags[ParquetValueType] = elementType
+			if elementConvertedType, exists := elementTags[ParquetConvertedType]; exists {
+				tags[ParquetValueConvertedType] = elementConvertedType
+			}
+		}
+	}
+	return
+}
+
+// Speical handling for the timestamp schema to convert it as millis as one int64 number instead of keeping it as struct.
```
Comment (samserpoosh) [inline schema/proto.go:249]: +1 on this ... Also, at this level, we don't have any notion of **timezone** I assume, right? Just wanna make sure that we'll be dealing with UTC all throughout the system to avoid issues and confusing errors 😄

File: schema/proto.go [hunk 22]
```diff
@@ -0,0 +1,386 @@
...
+	schema, err := common.NewSchemaElementFromTagMap(item.Info)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to generate schema for timestamp: %v", err)
+	}
+	newInfo := common.NewTag()
+	common.DeepCopy(item.Info, newInfo)
+	return schema, newInfo, nil
+}
+
+func NewSchemaHandlerFromProtoStruct(obj interface{}) (sh *SchemaHandler, err error) {
```
Comment (samserpoosh) [inline schema/proto.go:267]: Is `obj` going to be a **proto struct** passed into this method?

File: schema/proto.go [hunk 23]
```diff
@@ -0,0 +1,386 @@
...
+	}()
+
+	ot := reflect.TypeOf(obj)
+	if ot.Kind() == reflect.Pointer {
+		ot = ot.Elem()
+	}
+	item := NewItem()
+	item.GoType = ot
+	item.Info.InName = "Parquet_go_root"
+	item.Info.ExName = "parquet_go_root"
```
Comment (samserpoosh) [inline schema/proto.go:288]: Is this uppercase `P` vs lowercase `p` here intentional?

File: schema/proto.go [hunk 24]
```diff
@@ -0,0 +1,386 @@
...
+				if err != nil {
+					return nil, err
+				}
+				schemaElements = append(schemaElements, schema)
+				infos = append(infos, newInfo)
+				continue
+			}
+			schema := parquet.NewSchemaElement()
+			schema.Name = item.Info.InName
+			schema.RepetitionType = &item.Info.RepetitionType
```
Comment (samserpoosh) [inline schema/proto.go:314]: Is this `&` necessary? I see in other places when you're dealing with `item.Info`, you're not using the memory address via `&` operator 🤔

File: schema/proto.go [hunk 25]
```diff
@@ -0,0 +1,386 @@
...
+			for i := int(numField - 1); i >= 0; i-- {
+				f := item.GoType.Field(i)
+				if IsProtoInternalField(f.Name) {
+					continue
+				}
+				tagStr, err := GenerateFieldTag(f)
+				if err != nil {
+					return nil, err
+				}
+				newItem, err := generateSchemaForStructField(f, tagStr)
```
Comment (samserpoosh) [inline schema/proto.go:332]: At this point it's guaranteed that `f` is a `reflect.StructField`?

File: schema/proto.go (hunk 22 above)
Comment (samserpoosh) [inline schema/proto.go:267]: Given the conditionality, leveraging stack and other nuances involved in this method, I'd highly recommend a small paragraph description this method's behavior and what's trying to accomplish. Thanks!

File: schema/proto_test.go [hunk 26]
```diff
@@ -0,0 +1,128 @@
...
+	"github.com/stretchr/testify/assert"
+	"github.com/xitongsys/parquet-go/common"
+	"github.com/xitongsys/parquet-go/parquet"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+type JobStatus int32
+
+const (
+	JobStatus_JobStatus_UNSPECIFIED  JobStatus = 0
```
Comment (samserpoosh) [inline schema/proto_test.go]:
```suggestion
	JobStatus_UNSPECIFIED  JobStatus = 0
```

I assume?

File: schema/proto_test.go [hunk 27]
```diff
@@ -0,0 +1,128 @@
...
+
+func TestProtoSpecificSchema(t *testing.T) {
+	schemaHandler, err := NewSchemaHandlerFromProtoStruct(new(ProtoMessage))
+	if err != nil {
+		t.Errorf("failed to generate schema handler: %v", err)
+	}
+	assert.Equal(t, 3, len(schemaHandler.ValueColumns))
+	assert.Equal(t, parquet.Type_INT64, *schemaHandler.SchemaElements[1].Type)
+	assert.Equal(t, parquet.ConvertedType_TIMESTAMP_MILLIS, *schemaHandler.SchemaElements[1].ConvertedType)
+	assert.Equal(t, parquet.Type_BYTE_ARRAY, *schemaHandler.SchemaElements[2].Type)
```
Comment (samserpoosh) [inline schema/proto_test.go:61]: This is because given an Enum, we'll store the **string value** in Parquet, right? e.g. `JobStatus_ENQUEUED` as a string in the Parquet record.

File: schema/proto_test.go [hunk 28]
```diff
@@ -0,0 +1,128 @@
...
+	assert.Equal(t, parquet.ConvertedType_ENUM, *schemaHandler.SchemaElements[2].ConvertedType)
+	assert.Equal(t, parquet.Type_INT32, *schemaHandler.SchemaElements[3].Type)
+	if schemaHandler.SchemaElements[3].ConvertedType != nil {
```
Comment (samserpoosh) [inline schema/proto_test.go]:
```suggestion
	assert.Nil(schemaHandler.SchemaElements[3].ConvertedType)
```

Why not this ^^?

File: schema/schemahandler.go [hunk 29]
```diff
@@ -31,6 +32,30 @@ func NewPathMap(path string) *PathMapType {
...
+//
+//	Sex: Parquet_go_rootSex
+//	Name: Parquet_go_rootName
+//	Age: Parquet_go_rootAge
+//	Classes: Parquet_go_rootClasses
+//	  List: Parquet_go_rootClassesList
+//	    Element: Parquet_go_rootClassesListElement
+//	      Name: Parquet_go_rootClassesListElementName
+//	      Number: Parquet_go_rootClassesListElementNumber
+//	      Score: Parquet_go_rootClassesListElementScore
```
Comment (samserpoosh) [inline schema/schemahandler.go:47]: I **think** you're missing some spaces here. For instance, I assume you wanted:

```
Number: Parquet_go_root Classes List Element Number
```

As opposed to:

```
Number: Parquet_go_rootClassesListElementNumber
```

So you probably need to make some minor adjustments in the `output` method's implementation?

File: schema/schemahandler.go [hunk 30]
```diff
@@ -403,6 +340,109 @@ func NewSchemaHandlerFromStruct(obj interface{}) (sh *SchemaHandler, err error)
 	return res, nil
 }
 
+func generateSchemaForSlice(item *Item, stack []*Item, schemaElements []*parquet.SchemaElement, infos []*common.Tag) ([]*Item, []*parquet.SchemaElement, []*common.Tag) {
```
Comment (samserpoosh) [inline schema/schemahandler.go:343]: I don't see a dependency to `schemahandler.go` in `proto.go`. I assume it already exists? 😄

**Review by samserpoosh — COMMENTED** [review]
@wenruimeng-work It would be great to provide some high level documentation links or a small README with good examples to highlight this whole Proto => Parquet conversion business. And maybe even shine light on some tricky ones such as "Enum => string in Parquet along with its ConvertedType" in that READM 😄 

It would be super valuable IMHO to quickly glance at such a README and understand what this whole utility is trying to accomplish with a few well crafted examples in that doc. Let me know what you think please.

File: schema/proto.go (hunk 3 above)
Comment (wenrui-meng-ai) [reply schema/proto.go:76]: The byte array is basically string which is primitive type

File: schema/proto.go (hunk 5 above)
Comment (wenrui-meng-ai) [reply schema/proto.go]: type is reserved keyword, that's why I use tp. Maybe I can switch it typ

File: schema/proto.go (hunk 6 above)
Comment (wenrui-meng-ai) [reply schema/proto.go:146]: Yes

File: schema/proto.go (hunk 15 above)
Comment (wenrui-meng-ai) [reply schema/proto.go]: I thought I tested the interface variable will show Kind() as Interface, but it's directly returning the underlying type. I will just remove the Interface from here.

File: schema/proto.go (hunk 12 above)
Comment (wenrui-meng-ai) [reply schema/proto.go]: In go, if one of the case matched, it will skip all other branches and go the end. In this case, the end is just return statement. I don't there is much difference here. wdyt

File: schema/proto.go (hunk 9 above)
Comment (wenrui-meng-ai) [reply schema/proto.go:183]: The tag is used to store the actual primitive value. When it's struct, the schema generation will recursively traverse its fields.

File: schema/proto.go (hunk 11 above)
Comment (wenrui-meng-ai) [reply schema/proto.go:188]: Explained above for the err update. There is only one branch matched. Some branch will update err. return is just return tags, err.

File: schema/proto.go (hunk 14 above)
Comment (wenrui-meng-ai) [reply schema/proto.go:206]: Yes. In the test, we can see its format.

File: schema/proto.go (hunk 16 above)
Comment (wenrui-meng-ai) [reply schema/proto.go]: The convertedType might not need here. Let me remove it

File: schema/proto.go (hunk 18 above)
Comment (wenrui-meng-ai) [reply schema/proto.go:239]: Good point. I update it with bool to tell update key or value.

File: schema/proto.go (hunk 17 above)
Comment (wenrui-meng-ai) [reply schema/proto.go]: golang return without specifying is directly return the named return variable. Not sure whether it's recommended here but it's recommended in the GO programing book.

File: schema/proto.go (hunk 21 above)
Comment (wenrui-meng-ai) [reply schema/proto.go:249]: Yes. It's UTC for the proto Timestamp. The following is from proto lib. 
`
type Timestamp struct {
	state         protoimpl.MessageState
	sizeCache     protoimpl.SizeCache
	unknownFields protoimpl.UnknownFields

	// Represents seconds of UTC time since Unix epoch
	// 1970-01-01T00:00:00Z. Must be from 0001-01-01T00:00:00Z to
	// 9999-12-31T23:59:59Z inclusive.
	Seconds int64 `protobuf:"varint,1,opt,name=seconds,proto3" json:"seconds,omitempty"`
	// Non-negative fractions of a second at nanosecond resolution. Negative
	// second values with fractions must still have non-negative nanos values
	// that count forward in time. Must be from 0 to 999,999,999
	// inclusive.
	Nanos int32 `protobuf:"varint,2,opt,name=nanos,proto3" json:"nanos,omitempty"`
}
`

File: schema/proto.go (hunk 22 above)
Comment (wenrui-meng-ai) [reply schema/proto.go:267]: Yes

File: schema/proto.go (hunk 23 above)
Comment (wenrui-meng-ai) [reply schema/proto.go:288]: Yes. It distinguish the InName and ExName.

File: schema/proto.go (hunk 22 above)
Comment (wenrui-meng-ai) [reply schema/proto.go:267]: Updated.

File: schema/proto.go (hunk 24 above)
Comment (wenrui-meng-ai) [reply schema/proto.go:314]: Trying to not copy when it's pass around. That's what original repo implemented for each field.

File: schema/proto.go (hunk 25 above)
Comment (wenrui-meng-ai) [reply schema/proto.go:332]: Field() is always return StructField.

File: schema/proto_test.go (hunk 27 above)
Comment (wenrui-meng-ai) [reply schema/proto_test.go:61]: Yes

File: schema/schemahandler.go (hunk 29 above)
Comment (wenrui-meng-ai) [reply schema/schemahandler.go:47]: I didn't update its underlying path string. There is no space there, but it has a special character \x01 which is probably skipped in the print string. I don't think it's worth to split the path given the purpose only for us to understand the structure.

File: schema/schemahandler.go (hunk 30 above)
Comment (wenrui-meng-ai) [reply schema/schemahandler.go:343]: They are in the same package, which doesn't require import.

File: schema/schemahandler.go (hunk 29 above)
Comment (samserpoosh) [reply schema/schemahandler.go:47]: In case you don't think this would be hard to parse/inspect later down the line without the space, which would make it easier to quickly see what's going on (especially IF we'll be dealing with more complicated and nested data structures), then sounds good to me 😄

File: schema/schemahandler.go (hunk 30 above)
Comment (samserpoosh) [reply schema/schemahandler.go:343]: Got it, makes sense. Thanks!

File: schema/proto.go [hunk 31]
```diff
@@ -124,10 +126,12 @@ func IsProtoInternalField(name string) bool {
 	}
 }
 
-func GetDefinedTypeTag(tp reflect.Type) (tags map[string]string, err error) {
+// Get the tags from given reflect type through switch case, which has better performance than map.
+// Given this function is per event call, the performance matters when there are many events to process.
```
Comment (samserpoosh) [inline schema/proto.go:130]: Thanks, this context is good to have here IMHO 😄

File: schema/proto.go (hunk 9 above)
Comment (samserpoosh) [reply schema/proto.go:183]: Nice ... It **might worth** adding this explanation in the comment to avoid confusion.

File: schema/proto.go (hunk 11 above)
Comment (samserpoosh) [reply schema/proto.go:188]: So `return` in actuality becomes `return tags, err` behind the scene? 😮

File: schema/proto.go [hunk 32]
```diff
@@ -201,54 +205,40 @@ func getTagStringFromTags(tags map[string]string) string {
...
+		return nil, err
 	}
-	if IsPrimitiveOrPointerGoKind(tp.Elem().Kind()) {
-		elementTags, elementErr := GetDefinedTypeTag(tp.Elem())
+	return updateKeyOrValueType(tags, typ, false)
+}
+
+func updateKeyOrValueType(tags map[string]string, typ reflect.Type, updateKey bool) (map[string]string, error) {
+	if IsPrimitiveOrPointerGoKind(typ.Elem().Kind()) {
+		elementTags, elementErr := GetDefinedTypeTag(typ.Elem())
```
Comment (samserpoosh) [inline schema/proto.go]: Are you sure this is accurate now? Previously you had both `tp.Key().Kind()` and `tp.Elem().Kind()` in the logic and accordingly you captured their types via:

```go
...
elementTags, elementErr := GetDefinedTypeTag(tp.Key())
...
elementTags, elementErr := GetDefinedTypeTag(tp.Elem())
...
```

Now it seems like you're **only** doing this with `tp.Elem()`. Am I missing something?

File: schema/proto.go [hunk 33]
```diff
@@ -264,6 +254,11 @@ func generateSchemaTimestamp(item *Item) (*parquet.SchemaElement, *common.Tag, e
 	return schema, newInfo, nil
 }
 
+// DFS traverse the obj underlying struct type. If field is strucd recursive visit its sub-fields.
+// Generate the schemas for each non-struct field and create SchemaHandler from schema list.
+// Primitive type is just generating one schema for it.
+// Slice has its own special handling it. https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#lists
```
Comment (samserpoosh) [inline schema/proto.go]:
```suggestion
// Slice has its own special handling: https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#lists
```

File: schema/proto.go [hunk 34]
```diff
@@ -264,6 +254,11 @@ func generateSchemaTimestamp(item *Item) (*parquet.SchemaElement, *common.Tag, e
...
+// Map also requires special handling. https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#maps
```
Comment (samserpoosh) [inline schema/proto.go]:
```suggestion
// Map also requires special handling: https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#maps
```

File: schema/proto.go [hunk 35]
```diff
@@ -264,6 +254,11 @@ func generateSchemaTimestamp(item *Item) (*parquet.SchemaElement, *common.Tag, e
 	return schema, newInfo, nil
 }
 
+// DFS traverse the obj underlying struct type. If field is strucd recursive visit its sub-fields.
```
Comment (samserpoosh) [inline schema/proto.go:262]: I noticed you're explicitly using a `stack` in your DFS as opposed to leveraging **recursion**. Is this due to some performance concerns or ending up with **too deep call-stacks** since a Struct can be heavily nested at times?

**Review by samserpoosh — APPROVED** [review]
Overall LGTM and thanks for addressing the comments 👍 I added a couple more comments/questions as well. Please reply/address those prior to merging. Thanks!

> [!NOTE]
> I highly recommend testing the **parquet => DeltaStreamer => Hudi Table** portion sooner rather than later. IF you can cook some example Proto events, we can feed them to this utility you have and output some **parquets** accordingly. Then we can try shaping the Hudi Table's AVRO schema based on the parquet schema and ingesting those parquets using DeltaStreamer. This way we quickly check on E2E ingestion feasibility as early as possible in the process.

File: schema/proto.go (hunk 11 above)
Comment (wenrui-meng-ai) [reply schema/proto.go:188]: Yes

File: schema/proto.go (hunk 32 above)
Comment (wenrui-meng-ai) [reply schema/proto.go]: Oh Good catch. Need to fix it

File: schema/proto.go (hunk 35 above)
Comment (wenrui-meng-ai) [reply schema/proto.go:262]: Yes. Performance is one thing. In addition, they have similar implementation for other flow.
